sdks/python/apache_beam/transforms/stats_test.py

```diff
@@ -41,13 +40,15 @@
 from apache_beam.transforms.display import DisplayData
 from apache_beam.transforms.display_test import DisplayDataItemMatcher
 from apache_beam.transforms.stats import ApproximateQuantilesCombineFn
+from apache_beam.transforms.stats import ApproximateUniqueCombineFn
 
 
 class ApproximateUniqueTest(unittest.TestCase):
-  """Unit tests for ApproximateUnique.Globally and ApproximateUnique.PerKey.
-  Hash() with Python3 is nondeterministic, so Approximation algorithm generates
-  different result each time and sometimes error rate is out of range, so add
-  retries for all tests who actually running approximation algorithm."""
+  """Unit tests for ApproximateUnique.Globally, ApproximateUnique.PerKey,
+  and ApproximateUniqueCombineFn.
```

Comment:
Can we parameterize the test to exercise both md5/mmh3 codepaths. We could force md5 by a patch for test purposes if mmh3 is installed.

Reply:
Here is a background of mmh3 in case it helps for decision making. I used mmh3 at first but decided to use default hash function. mmh3 has better prediction performance, but it introduces much more complexity for dataflow import, and users should use HLL if they really care about the accuracy. In addition, the default hash function improved for Py3 and the performance diff was not big enough to use mmh3 and introduce the complexity. This is the reason mmh3 was not used, and it is just for reference.  
I think it’s better to mention that HLL is supported by Beam.

Reply:
We can still do Dataflow import with a new dependency, AFAIK we noticed that Google internal-only tests for Beam on Windows were failing, so we thought that adding mmh3 would be a concern for all Beam Windows users. I no longer see Windows failures in precommit tests running on the PR.

Reply:
I agree with Hannah. HLL would be the right solution and given that py3 has a better hash function adding this dependency could be avoided.

What is the specific issue we are addressing by this change?

Reply:
hash is non deterministic, so if the computation is distributed over different machines, the estimates are overly inaccurate

Reply:
py3 hash is non-deterministic, see BEAM-10824 and internal issue b/166646014 which has a bit more details. 
We should not be using py3 hash function, and I support this change. Current change supports both md5 and mmh3, but we have a remanining question whether mmh3 should be a default dep (better performance than md5), but extra dependency, or optional (default to md5, give a warning recommendation to install mmh3).

HLL is not available for Beam Python yet, is it?

Reply:
I did not know about py3 hash being non deterministic. That is strange.

In that case I agree with this change. Let's make it optional so that the getting started experience does not change and any actual execution environment could install it as needed. (i.e. add it here : https://github.com/apache/beam/blob/master/sdks/python/container/base_image_requirements.txt in this PR.)

Reply:
sounds good thanks! more info on py3 hash here: https://stackoverflow.com/questions/27522626/hash-function-in-python-3-3-returns-different-results-between-sessions